API-GUIDELINE.md

```diff
@@ -0,0 +1,255 @@
+# API GUIDELINES
+
+## Overview
+As commercetools is 100% API First, it's important to have a well designed developer experience.
+Additionally we aim to achieve more automation and improve the review process of new changes in the APIs.
+
+With this in mind, we have created "API Guidelines" that aims to help developers to design APIs:
+
+ - simple
+ - consistent 
+ - self-explanatory
+ 
+The Keywords to have a good Guidelines are:
+ 
+ - Consistency => following the Guidelines
+ - Uniformity => checking in the code if there are other properties with similar names and adjusting field naming based on it
+ - Explicability => explaining in a few words the purpose of the property avoiding abbreviations (there are some [exceptions](#exceptions)).
+
+## Scope
+The purpose of these guidelines is to define consistent practices and patterns for all API endpoints.
+
+## URI
+Our URIs are composed by:
+   
+    Base Uri like:
+        https://{service}.{region}.{cloudProvider}.commercetools.com 
+
+    {projectKey}
+
+    resources
+
+    by ID
+
+    by external ID with equals e.g. customers/key={key}
+
+See our documentation for more details https://docs.commercetools.com/http-api.html#hosts and  https://docs.commercetools.com/http-api-authorization#requesting-an-access-token-using-commercetools-oauth2-server 
+
+## Resources
+A Resource is a data collection object or a data representation of an entity in the e-commerce domain, which is specified by properties and methods to operate on them.
+
+A resource can be an entity or a collection:
+ - Entity is a resource that has to be in singular.
+ - Collections are a set of resources that must be in plural.
```

Comment:
Please ignore this comment if it is clear to developers (after all I think this is an internal document for technical people).

To me " ... are a set ..." sounds suspicious. This was the starting point.

I am not familiar with the technical things, so I am afraid of making unnecessary noise here, but still I do not understand lines 41-43. The lines 39 and 41 explains a resource in terms of an entity and a collection, where as the lines 42 and 43 explains an entity and collections in terms of resources. (eg. in l. 39, A Resource is ... or a data rep. of an entity; in l. 42, Entity is a resource ... )

Also, does the line 43 mean "A collection is a set of more than one resources"? If so, it contradicts the line 41, which in particular says that *a* resource can be a collection.

I am afraid that I simply missed some intended difference between **R**esource (l. 39) and **r**esource (else where) or I simply do not know the technical definition of the term "be in singular/plural". If this is the case, I am sorry.

Reply:
no problem Tohru, here the clarification in the line 39 there is the definition easily you can say that there is simple resource (it's just one result) = Entity or a collection resource(it's more then one result) = Collection.

Reply:
Thank you! Now I understand that the line 39 is the definition of a **Resource** in terms of data collection object and an entity. The line 39 itself is now clear. 

With this in mind, the line 41 seems to be either a rephrase of the line 39 or an explanation of the behavior of a single **R**esource. 

Either way, the lines 42 and 43 try to explain the words "entity" and "collection" used in the line 41, but it is done so in terms of Resources, which was defined in terms of entity and collection.
 
I am not sure but I am afraid that this is a circular argument or some word (collection, entity or R/resource) is used in more than one meaning.

So, I think there is a room for improvement here. But since I do not know what you mean to say, I can not unfortunately offer a suggestion. If the passage is clear to others, I am happy to let it go, but if not, maybe we can rephrase a bit?

Reply:
ok, I will think about it, thanks!

Reply:
Just one final suggestion. If this is about the naming convention, you might want to write so more explicitly and perhaps with examples. 

An entity as a reference must be named in singular (eg. ... ).
A collection as a reference must be named in plural (eg. ... ).